pkg/geo/wkt/lex.go

```diff
@@ -275,43 +313,164 @@ func (l *wktLex) isValidStrideForLayout(stride int) bool {
+		if l.getCurLayout() == geom.XYM {
+			l.setCurLayoutMustBeEmpty(true)
+		}
+		return true
+	}
+
+	if l.getCurLayout() == geom.XYM {
+		l.setIncorrectLayoutError(geom.NoLayout, "XYM layout must use the M variant of the geometry type")
+		return false
+	}
+
 	return true
 }
 
-func (l *wktLex) setLayoutEmptyInCollection() bool {
-	if l.curLayout == geom.XY {
+func (l *wktLex) setLayoutBaseTypeEmpty() bool {
+	if !l.getCurLayoutBaseType() {
+		// Handle edge case where a GEOMETRYCOLLECTIONM may have base type EMPTY geometries but not non-EMPTY geometries.
+		if l.getCurLayout() == geom.XYM {
+			l.setCurLayoutMustBeEmpty(false)
+		}
 		return true
 	}
-	if l.curLayout == geom.NoLayout {
-		l.curLayout = geom.XY
+
+	switch curLayout := l.getCurLayout(); curLayout {
+	case geom.NoLayout:
+		l.setLayout(geom.XY)
+		fallthrough
+	case geom.XY:
 		return true
+	default:
+		l.setIncorrectLayoutError(geom.XY, "EMPTY is XY layout in base geometry type")
+		return false
 	}
-	l.setIncorrectLayoutError(geom.XY, "EMPTY is XY layout in base geometry type collection")
-	return false
 }
 
 func (l *wktLex) setLayoutIfNoLayout(layout geom.Layout) {
-	switch l.curLayout {
+	switch curLayout := l.getCurLayout(); curLayout {
 	case geom.NoLayout:
-		l.curLayout = layout
+		l.setCurLayout(layout)
 	case geom.XY, geom.XYM, geom.XYZ, geom.XYZM:
 		break
 	default:
 		// This should never happen.
-		panic("unknown geom.Layout")
+		panic(fmt.Sprintf("unknown geom.Layout %d", layout))
+	}
+}
+
+func (l *wktLex) pushLayoutStack(layout geom.Layout) bool {
+	// Check that the new layout is compatible with the previous ones.
+	// baseType inherits from outer context.
+	newStackObj := layoutStackObj{
+		layout:   layout,
+		baseType: l.getCurLayoutBaseType(),
+	}
+
+	switch layout {
+	case geom.NoLayout:
+		newStackObj.layout = l.getCurLayout()
+	case geom.XYM, geom.XYZ, geom.XYZM:
+		if layout != l.getCurLayout() && l.getCurLayout() != geom.NoLayout {
+			l.setIncorrectLayoutError(layout, "")
+			return false
+		}
+		newStackObj.baseType = false
+	case geom.XY:
+		// This should never happen. Base type should be pushing geom.NoLayout.
+		fallthrough
+	default:
+		// This should never happen.
+		panic(fmt.Sprintf("unknown geom.Layout %d", layout))
+	}
+
+	l.layoutStack = append(l.layoutStack, newStackObj)
+	return true
+}
+
+func (l *wktLex) popLayoutStack() geom.Layout {
+	top := l.getCurLayout()
+	l.layoutStack = l.layoutStack[:len(l.layoutStack)-1]
+	// Update the outer context with the type we parsed in the inner context.
+	ok := l.setLayout(top)
+	if !ok {
+		// This should never happen. Any layout incompatibility should error at the point it's discovered.
+		panic("uncaught layout incompatibility")
+	}
+	return top
+}
+
+func (l *wktLex) topLayoutStack() layoutStackObj {
+	l.checkLayoutStackNotEmpty()
+	return l.layoutStack[len(l.layoutStack)-1]
+}
+
+func (l *wktLex) getCurLayout() geom.Layout {
```

Comment:
nit: it's usually not Go-matic to use `get`. omit `get` - `curLayout()`, `curLayoutBaseType()`, `curLayoutMustBeEmpty`.

Note you may see it around if the non-getter name clashes.

See: https://golang.org/doc/effective_go#Getters

Reply:
Oh I see. Does this apply even if the field isn't actually called `curLayout`?